pkg/controller/appdefinition/networkpolicy.go

```diff
@@ -1,25 +1,257 @@
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      app.Name,
-			Namespace: app.Status.Namespace,
+			Namespace: podNamespace,
+		},
+		Spec: networkingv1.NetworkPolicySpec{
+			PodSelector: metav1.LabelSelector{
+				MatchLabels: labels.Managed(app),
+			},
+			Ingress: []networkingv1.NetworkPolicyIngressRule{{
+				From: allowedNamespaceSelectors,
+			}},
+			PolicyTypes: []networkingv1.PolicyType{networkingv1.PolicyTypeIngress},
+		},
+	})
+	return nil
+}
+
+// NetworkPolicyForIngress creates Kubernetes NetworkPolicies to allow traffic to exposed HTTP ports on
+// Acorn apps from the ingress controller. If the ingress controller namespace is not defined, traffic from
+// all namespaces will be allowed instead.
+func NetworkPolicyForIngress(req router.Request, resp router.Response) error {
+	cfg, err := config.Get(req.Ctx, req.Client)
+	if err != nil {
+		return err
+	} else if !*cfg.NetworkPolicies {
+		return nil
+	}
+
+	ingress := req.Object.(*networkingv1.Ingress)
+	appName := ingress.Labels[labels.AcornAppName]
+	projectName := ingress.Labels[labels.AcornAppNamespace]
+
+	// create a mapping of k8s Service names to published port names/numbers
+	svcNameToPorts := make(map[string][]networkingv1.ServiceBackendPort)
+	for _, rule := range ingress.Spec.Rules {
+		for _, path := range rule.HTTP.Paths {
+			svcName := path.Backend.Service.Name
+			port := path.Backend.Service.Port
+			svcNameToPorts[svcName] = append(svcNameToPorts[svcName], port)
+		}
+	}
+
+	for svcName, ports := range svcNameToPorts {
+		// get the Service from k8s
+		svc := corev1.Service{}
+		err = req.Get(&svc, ingress.Namespace, svcName)
+		if err != nil {
+			// service doesn't exist yet, so retry in 3 seconds
+			resp.RetryAfter(3 * time.Second)
+			return nil
+		}
+
+		netPolName := name.SafeConcatName(projectName, appName, ingress.Name, svcName)
+
+		// build the namespaceSelector for the NetPol
+		var namespaceSelector metav1.LabelSelector
+		if *cfg.IngressControllerNamespace != "" {
+			namespaceSelector = metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"kubernetes.io/metadata.name": *cfg.IngressControllerNamespace,
+				},
+			}
+		}
+
+		// build the port slice for the NetPol
+		var netPolPorts []networkingv1.NetworkPolicyPort
+		for _, port := range ports {
+			// try to map this ingress port to a port on the service
+			for _, svcPort := range svc.Spec.Ports {
+				if (svcPort.Name != "" && svcPort.Name == port.Name) || svcPort.Port == port.Number {
+					targetPort := svcPort.TargetPort
+					netPolPorts = append(netPolPorts, networkingv1.NetworkPolicyPort{
+						Protocol: &[]corev1.Protocol{corev1.ProtocolTCP}[0],
+						Port:     &targetPort,
+					})
+					netPolName = name.SafeConcatName(netPolName, strconv.Itoa(int(targetPort.IntVal)))
+				}
+			}
+		}
+
+		if len(netPolPorts) == 0 {
+			logrus.Warnf("found no matching ports between Ingress %s and Service %s in Namespace %s", ingress.Name, svcName, ingress.Namespace)
```

Comment:
Is this really a warn level statement? That feels a bit severe

Reply:
Yeah I wasn't really sure. Theoretically this should never happen - if it does, there's a logic error somewhere else in the Acorn controller. So I wasn't sure whether it would be appropriate for this handler to return an error in this case. But I can make it return an error if that would be better.